lib/Sema/TypeCheckAttr.cpp

```diff
@@ -3008,8 +3008,32 @@ void AttributeChecker::visitPropertyWrapperAttr(PropertyWrapperAttr *attr) {
 }
 
 void AttributeChecker::visitFunctionBuilderAttr(FunctionBuilderAttr *attr) {
-  // TODO: check that the type at least provides a `sequence` factory?
-  // Any other validation?
+  auto *nominal = dyn_cast<NominalTypeDecl>(D);
+  SmallVector<ValueDecl *, 4> potentialMatches;
+  bool supportsBuildBlock = TypeChecker::typeSupportsBuilderOp(
+      nominal->getDeclaredType(), nominal, D->getASTContext().Id_buildBlock,
+      /*argLabels=*/{}, &potentialMatches);
+
+  if (!supportsBuildBlock) {
+    diagnose(nominal->getLoc(), diag::function_builder_static_buildblock);
+
+    // For any close matches, attempt to explain to the user why they aren't
+    // valid.
+    for (auto *member : potentialMatches) {
+      if (member->isStatic() && isa<FuncDecl>(member))
+        continue;
+
+      if (isa<FuncDecl>(member) &&
+          member->getDeclContext()->getSelfNominalTypeDecl() == nominal)
```

Comment:
I don't think you need the "nominal" check here; we've generalized the lookup so `buildBlock` can come from, e.g., a superclass or conformed-to protocol.

Reply:
@DougGregor The goal for that check was to avoid offering the fix-it to add “static “ to a `buildBlock` declaration which is not part of the type that was tagged with `@_functionBuilder` since it’s less likely that the user intended for that to be a match (and that declaration may be in use by other non-function-builder types). Declarations that are from e.g. a superclass or protocol will still get the “potential match is not a static method” diagnostic note, just not the fix-it. Happy to change this if you think we should offer the fix-it wherever the instance `buildBlock` is located!

Reply:
Ahh, I understand now. Thank you for clarifying!